### PR TITLE
feat(api): thermocycler-gen2 firmware update

### PR DESCRIPTION
**Pull request by fsinapi**
# Overview
Implements changes required to upload firmware updates to a Thermocycler Gen2.

# Changelog

- Removed hard-coded module firmware prefixes from the module abstract base class, and added a dedicated function to get the firmware file prefix for each module.
- Thermocycler can now handle the differences needed to upload firmware to a Gen2 vs a Gen1, including finding the right file
- The function to find a DFU device through `dfu-util` now accepts a parameter for the number of unique "devices" to expect, because this varies with the exact microcontroller. The F303 on the Heater-Shaker has two endpoints, while the Thermocycler-Gen2's G491 has three due to the one-time-programmable memory
- Fixed an issue with the Thermocycler Gen2 function to enter the bootloader where it would throw an exception when trying to configure the USB port after the Thermocycler had already shut down

Tested the following with a robot that had the v0.5.2 firmware saved in `/usr/lib/firmware`:
- Connected a Thermocycler Gen2 with 0.5.1, used the app to update firmware. Thermocycler rebooted and the version number was now 0.5.2, and the app did not show that an update was needed
- Changed name of the firmware file to 0.5.3 and updated again. Thermocycler rebooted still on 0.5.2 and the app still said a new update was available
- Connected a Thermocycler Gen1 with 1.1.0, the app did not say an update was available
- Changed the name of the Gen2 firmware file to version 1.2.0, the Thermocycler Gen1 still did not show an update available

# Review requests

- Confirm that this approach of getting the name of the update file from the instance of a module makes sense

# Risk assessment

Medium/low, there are changes to the code flow when updating all modules but the changes are simple and well-tested.

### CODE REVIEW

**Review by sfoster1 — APPROVED** [review]
Probably worth noting why there's two or three devices in the comment next to the device count but overall looks great!

File: api/src/opentrons/hardware_control/modules/mod_abc.py [hunk 1]
```diff
@@ -73,13 +73,7 @@ def get_bundled_fw(self) -> Optional[BundledFirmware]:
...
         if not IS_ROBOT:
             return None
-        name_to_fw_file_prefix = {
-            "tempdeck": "temperature-module",
-            "magdeck": "magnetic-module",
-            "heatershaker": "heater-shaker",
-        }
-        name = self.name()
-        file_prefix = name_to_fw_file_prefix.get(name, name)
+        file_prefix = self.firmware_prefix()
```
Comment (Laura-Danielle) [inline api/src/opentrons/hardware_control/modules/mod_abc.py:76]: aren't some of the older firmware files called `tempdeck` and `magdeck` for the arduinos?

**Review by Laura-Danielle — APPROVED** [review]
I think it's fine to get the name of the update file from the module instance. I did have a question about the prefixes though. I had thought we were still using `tempdeck` and `magdeck` somewhere in the firmware update.

**Review by vegano1 — APPROVED** [review]

File: api/src/opentrons/hardware_control/modules/mod_abc.py (hunk 1 above)
Comment (fsinapi) [reply api/src/opentrons/hardware_control/modules/mod_abc.py:76]: Was the old version of this function able to find those, though? The way this was set up with the `name_to_fw_file_prefix` would _only_ match a firmware file if it started with the `temperature-module` or `magnetic-module` names, the `self.name()` was only used for lookup if there isn't an entry in `name_to_fw_file_prefix`

FWIW buildroot packages everything under the latter names as well.

File: api/src/opentrons/drivers/thermocycler/driver.py [hunk 2]
```diff
@@ -332,7 +332,5 @@ async def enter_programming_mode(self) -> None:
             gcode=GCODE.ENTER_PROGRAMMING
         )
         # The timeout is overwritten because no response should ever come
```
Comment (sanni-t) [inline api/src/opentrons/drivers/thermocycler/driver.py]: This comment can be removed

File: api/src/opentrons/hardware_control/modules/mod_abc.py (hunk 1 above)
Comment (sanni-t) [reply api/src/opentrons/hardware_control/modules/mod_abc.py:76]: Nice! I like this change

**Review by sanni-t — APPROVED** [review]
Looks great! 
As a sanity check, I'd recommend testing a gen1 thermocycler update as well on the robot (and maybe even tempdeck and magdeck updates) so that we're not surprised by any unexpected behavior change later.

**Comment by fsinapi** [comment]
> As a sanity check, I'd recommend testing a gen1 thermocycler update as well on the robot (and maybe even tempdeck and magdeck updates) so that we're not surprised by any unexpected behavior change later.

Tested the above, those workflows look good to go 👍

**Pull request merged**